vendor/github.com/orisano/fls/list_linux.go

```diff
@@ -0,0 +1,206 @@
+// +build linux
+
+// Copyright 2009 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//    * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//    * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//    * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package fls
+
+import (
+	"os"
+	"path"
+	"runtime"
+	"syscall"
+	"unsafe"
+)
+
+func ListDirectories(dirpath string, parent string, recursive bool, output map[string]struct{}) error {
+	f, err := os.Open(dirpath)
+	if err != nil {
+		// Ignore if this hierarchy does not exist.
+		if os.IsNotExist(err) {
+			err = nil
+		}
+		return err
+	}
+
+	fd := int(f.Fd())
+	buf := make([]byte, 8192)
+	var nbuf, bufp int
+
+	dirs := make([]string, 0, 128)
+	for {
+		if bufp >= nbuf {
+			bufp = 0
+			var errno error
+			nbuf, errno = syscall.ReadDirent(fd, buf)
+			runtime.KeepAlive(f)
+			if errno != nil {
+				return os.NewSyscallError("readdirs", errno)
+			}
+			if nbuf <= 0 {
+				break
+			}
+		}
+		var nb int
+		nb, dirs = parseDirent(buf[bufp:nbuf], dirs)
+		bufp += nb
+	}
+	f.Close()
+
+	for _, d := range dirs {
+		name := path.Join(parent, d)
+		output[name] = struct{}{}
+
+		// List subcontainers if asked to.
+		if recursive {
+			err := ListDirectories(path.Join(dirpath, d), name, true, output)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func parseDirent(buf []byte, dirs []string) (consumed int, newdirs []string) {
```

Comment:
it looks like we could just use the golang implementation directly, since it is public? https://github.com/golang/go/blob/master/src/syscall/dirent.go#L64

Reply:
parseDirent is edited for handle Dirent.Type.
original implementation can't handle Dirent.Type.